support/arr/arr.go

```diff
@@ -0,0 +1,633 @@
+
+			res = append(res, val...)
+		}
+	}
+
+	return res
+}
+
+// Forget Remove one or many array items from a given array.
+func Forget[T any](arr []T, keys any) ([]T, error) {
+	if len(arr) == 0 || keys == nil {
+		return arr, nil
+	}
+
+	switch v := keys.(type) {
+	case int:
+		keys = []int{v}
+	case []int:
+		sort.Sort(sort.Reverse(sort.IntSlice(v)))
+	default:
+		return arr, ErrInvalidKeys
+	}
+
+	for _, v := range keys.([]int) {
+		if v >= 0 && v < len(arr) {
+			copy(arr[v:], arr[v+1:])
+			arr[len(arr)-1] = reflect.Zero(reflect.TypeOf(arr[0])).Interface().(T)
+			arr = arr[:len(arr)-1]
+
+			continue
+		}
+	}
+	return arr, nil
+}
+
+// Get an item from an array using int key.
+// todo: thread safe?
+func Get[T any](arr []T, key int, def any) (T, bool) {
+	if key < 0 || key > len(arr)-1 {
+		return def.(T), false
+	}
+
+	return arr[key], true
+}
+
+// Has checks if an item or items exist in an array using "dot" notation.
+func Has[T any](arr []T, keys any) bool {
+	if len(arr) == 0 || keys == nil {
+		return false
+	}
+
+	switch v := keys.(type) {
+	case int:
+		keys = []int{v}
+	case []int:
+	default:
+		return false
+	}
+
+	for _, v := range keys.([]int) {
+		if v < 0 || v >= len(arr) {
+			return false
+		}
+	}
+
+	return true
+}
+
+// HasAny Determine if any of the keys exist in an array using int key
+func HasAny[T any](arr []T, keys any) bool {
+	if len(arr) == 0 || keys == nil {
+		return false
+	}
+
+	switch v := keys.(type) {
+	case int:
+		keys = []int{v}
+	case []int:
+	default:
+		return false
+	}
+
+	for _, v := range keys.([]int) {
+		if v >= 0 && v < len(arr) {
+			return true
+		}
+	}
+
+	return false
+}
+
+// IsAssoc Determines if an array is associative.
+func IsAssoc[T any](arr T) bool {
+	k := reflect.ValueOf(arr).Kind()
+	return k == reflect.Map
+}
+
+// IsList Determines if an array is a list.
+func IsList[T any](arr T) bool {
+	k := reflect.ValueOf(arr).Kind()
+	return k == reflect.Slice || k == reflect.Array
+}
+
+// Join concatenates elements of a slice into a string with a specified delimiter and final separator
+func Join[T any](arr []T, delimiter string, finalSeparator ...string) string {
+	l := len(arr)
+	if l == 0 {
+		return ""
+	}
+	if l == 1 {
+		return fmt.Sprint(arr[0])
+	}
+	if l == 2 {
+		return fmt.Sprintf("%v%s%v", arr[0], finalSeparator[0], arr[1])
+	}
+	if len(finalSeparator) == 0 {
+		finalSeparator = []string{delimiter}
+	}
+
+	var builder strings.Builder
+	for i, v := range arr {
+		builder.WriteString(fmt.Sprint(v))
+		if i == len(arr)-2 {
+			builder.WriteString(finalSeparator[0])
+		} else if i < len(arr)-1 {
+			builder.WriteString(delimiter)
+		}
+	}
+	return builder.String()
+}
+
+// KeyBy Key an associative array by a field or using a callback.
+func KeyBy[T any, K comparable](array []T, key K) ([]T, error) {
+	return nil, ErrNoImplementation
+}
+
+// PrependKeysWith Prepend the key names of an associative array.
+func PrependKeysWith[T any](array []T, prependWith string) ([]T, error) {
```

Comment:
S: We can remove it, implement it in map.

Reply:
done